### PR TITLE
[FE] Makka / #1992 pay with bank transfer FE changes

### LINKED ISSUE #1992: [retire] Add pay with bank transfer option - user journey priority #1
### Related docs & links
[shipped figma designs](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=1830-17766&mode=design&t=jZBOtWJqyzfOTFSh-0)

[Design/discovery discord thread](https://discord.com/channels/1135827610434289727/1176908485368233985)

### Requirements & user journey

- User journey starts from a [project details page](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=1830-18024&mode=design&t=jZBOtWJqyzfOTFSh-4)
- User clicks RETIRE NOW on a pool or seller listing (support for this is in progress in #1659 )
- This opens the [retire payment form](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=1810-16827&mode=design&t=jZBOtWJqyzfOTFSh-4)
- Need to add a Bank Transfer Pay with: option
Screenshot 2023-12-07 at 9 13 26 AM

- User clicks Bank Transfer Pay with: option
- User fills in the required "How many tonnes..." field and in this case has also filled in beneficiary name and retirement message (although beneficiary name, beneficiary wallet, and retirement message are optional fields, same as when paying with USDC). This [activates the RETIRE CARBON button which the user clicks](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=1810-23803&mode=design&t=jZBOtWJqyzfOTFSh-4)
- User is next presented with a [Confirm Purchase modal](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=1810-24185&mode=design&t=jZBOtWJqyzfOTFSh-4); they click CONTINUE (Note...GO BACK returns the user to the retire payment form)
- User completes the [Pay with bank transfer form](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=1810-24580&mode=design&t=jZBOtWJqyzfOTFSh-4)

**Pay with bank transfer form notes**
- LEARN MORE will go to a marketing blog post (still to be written, URL to be provided)
- The project, tonne quantity, and any optional fields filled in on the retire payment form needs to be passed to the Pay with bank transfer form. We do not want to have to make the user fill out the information twice.
- Note the form differences between desktop and mobile
- Note required form fields...you will need to add some error handling here using our standard of highlighting the field in a red box and providing helper text

- User completes the form with no required field errors and clicks the SUBMIT button
- ~On SUBMIT, the form contents should be send to **sales email address** (to be provided)~
- Rather than an email, we will write the form data to hubspot using their api, see comment below
- Once that process completes, show the user a [Form complete thank you page](https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=1830-26135&mode=design&t=jZBOtWJqyzfOTFSh-4)
- The Form complete page should show all the form fields and data that were submitted

- This completes the pay with bank transfer user journey. Clicking BROWSE PROJECTS on the Form complete page returns the user to the Marketplace.

### PR DESCRIPTION
**Pull request by 0xMakka**
## Description

This PR adds the ability to Retire via bank transfer, more detailed descriptions are available in the main ticket - #1992  

## Related Ticket

Closes #1992
Also related to & relies on #2024 

## Notes For QA

QA is required as this PR adds a new feature to pay with bank transfer on retirement pools.
* [x] Allow a user to choose Bank transfer from the `Pay with` list.
* [x] `How many tonnes of carbon would you like to retire?` is the only required field at this stage - the "Retire Carbon" button should be disabled if nothing is entered in this field. The other fields are optional. If nothing is entered into the Beneficiary wallet address field, the connected users wallet address will be used.
* [x] After entering a quantity and clicking "Retire Carbon", the user should be presented with a modal popup according to the figma designs where they can "Continue" or "Go back". 
      - Upon clicking "Go back" the modal should be closed.
      - Upon clicking "Continue", the user should be brought to the "Pay with Bank Transfer" form where they are presented with a more detailed form with both required and optional fields. Any details entered on the previous "Retire Carbon" form should be carried over and filled in on this form.
* [x] On the "Pay with Bank Transfer" form the "Submit" button should be disabled if the required fields aren't filled out (denoted with an * on the label). As the user types in the input fields, error messages should be shown if the data is incorrect on a required field.
* [x] Upon submitting the form, a POST request is made to the carbonmark api at the following endpoint `/retire/bank-transfer` with the form data in the body of the request. The api then makes a call to the hubspot form api with the form data. 
     - If the request is successful, a 200 response will be returned and the user will be shown the "Form complete" screen with the previously entered data filled out.
     - If the request is unsuccessful, a 400 response will be returned and an error message will be displayed above the "Submit" button.

Specific pages, components or journeys that might be affected:
- Pool Retirements Page

### CODE REVIEW

File: carbonmark/components/pages/Project/Retire/Pool/RetireInputs.tsx [hunk 1]
```diff
@@ -86,6 +86,30 @@ const validations = (
       },
     },
   },
+  "bank-transfer": {
+    quantity: {
+      min: {
+        // @todo - makka - confirm with atmos/brice what the min quantity is
```
Comment (0xMakka) [inline carbonmark/components/pages/Project/Retire/Pool/RetireInputs.tsx:92]: @jabby09 is there a minimum quantity for bank transfers? I've currently set this to the same as credit card.

**Comment by jabby09** [comment]
@0xtapi i will be reviewing this one as well
designs are here: https://www.figma.com/file/5GewbdVaM1Q06CwHmF7d3F/Projects?type=design&node-id=2013-22744&mode=design&t=S61kzPdFZAf3aXPd-4

can you please focus on mobile and form field error checking

not sure you have access to hubspot?

**Review by jabby09 — CHANGES_REQUESTED** [review]
### UPDATE - RESOLVED

@0xMakka just started QA process
first "happy path" was a success!

one quick thing I noticed, in the project section of the form we should also include the vsc number and vintage, so ideally include VCS-981-2017 in the project name as an example, maybe in () after the name?

**Comment by 0xtapi** [comment]
> can you please focus on mobile and form field error checking
> 

Will do!

> not sure you have access to hubspot?

I do not.

**Comment by 0xtapi** [comment]
On mobile (but probably same on desktop):
- While filling out the personal details form, app doesn't check whether the email address is valid. Any value (like "abc123") will be accepted.
- "First name" and "Last name" allow numbers in the names, so user could enter something like "Max123" and it would be accepted by the app. Are there any regions where numbers are part of the name...? If not, this should probably check whether it's a "clean" name without numbers.

> is there a maximum total price for bank transfers? I've currently set this to the same as credit card - 2,000.

In the "How would would you like to retire?" field, the app currently states it must be at least one tonne. This should be changed to whatever the minimum value is (once we know this value). Minor nitpick: we should maybe change this question to "How many tonnes would you like to retire?" to make it clear we're asking for tonnes and not for dollar values or so.

Btw, this whole process looks very good on mobile. No page breakers, bad layout, wrong font sizes or anything of that sort. Sweet stuff!

**Comment by jabby09** [comment]
> n the "How much would you like to retire?" field, the app currently states it must be at least one tonne. This should be changed to whatever the minimum value is (once we know this value). Minor nitpick: we should maybe change this question to "How many tonnes would you like to retire?" to make it clear we're asking for tonnes and not for dollar values or so.

@0xMakka @0xtapi hmmm...I did not see this enforced on the CM form but I tried to do quantity of 5000 and nothing happens, I just click RETIRE CARBON and nothing (and no error messages)

That being said, for Bank Transfer there would be no maximum quantity so if that is being enforced it should be removed.

I can ask the sales guys if they want a minimum value, stay tuned.

**Review by jabby09 — CHANGES_REQUESTED** [review]
@0xMakka 

Updates from the business team.

- [ ] Change `1 business day` to `2 business days` on the confirm purchase modal and Pay with bank transfer form intro text. 
- [ ] In the CM form, when a user selects Pay with Bank transfer, on the tonnes Quantity field, there is no maximum, however, the minimum is 100 tonnes.

**Comment by 0xtapi** [comment]
Almost there, but the check for emails still isn't quite working. It seems to check whether an "@" is there, but nothing else. So I can enter something like "tapioka@hello" or really wild stuff like "hi@-" and it goes through. Need a more strict check here (something like "Is this string@string.tld?" if not -> error).

All the other points I raised have been fixed though!

**Review by jabby09 — CHANGES_REQUESTED** [review]
@0xMakka another minor copy changes requested from the business team.

On the Confirm Purchase and Form copy change:

...`within 2 business days`...

to:

...`within 2 business days (upon receiving all required information)`...

**Comment by 0xMakka** [comment]
> Almost there, but the check for emails still isn't quite working. It seems to check whether an "@" is there, but nothing else. So I can enter something like "tapioka@hello" or really wild stuff like "hi@-" and it goes through. Need a more strict check here (something like "Is this [string@string.tld](mailto:string@string.tld)?" if not -> error).
> 
> All the other points I raised have been fixed though!

@0xtapi yeah you are right, I initially only went with the "@" to not make it too strict. However, I've just updated to be a little bit stricter regex which requires both an "@" a "." to be present in the string. So it should catch the cases you mentioned above.

**Comment by 0xtapi** [comment]
> @0xtapi yeah you are right, I initially only went with the "@" to not make it too strict. However, I've just updated to be a little bit stricter regex which requires both an "@" a "." to be present in the string. So it should catch the cases you mentioned above.

Lookss good now!

**Review by jabby09 — CHANGES_REQUESTED** [review]
@0xMakka unfortunately we do not have a blog article yet to link to and probably not until next year.

So we need to remove / hide the LEARN MORE button. Probably also mean shifting up the form to maintain proper page spacing.

image

**Review by jabby09 — APPROVED** [review]

**Review by Atmosfearful — CHANGES_REQUESTED** [review]
I don't want this in the public API docs confusing application developers since it's just form data specific to our frontend.
This would be better served from carbonmark/pages/api-- should be able to copy paste the core logic over, minus the fastify/schema and test boilerplate.

image

You can bring the test over but we don't have a test runner for `carbonmark` yet.

https://nextjs.org/docs/pages/building-your-application/routing/api-routes

**Comment by 0xMakka** [comment]
> I don't want this in the public API docs confusing application developers since it's just form data specific to our frontend.
> 
> image
> 
> This would be better served from carbonmark/pages/api-- should be able to copy paste the core logic over directly [(see nextjs docs)](https://nextjs.org/docs/pages/building-your-application/routing/api-routes), minus the fastify/schema and test boilerplate.
> 
> We don't have a test runner for `carbonmark` yet but a simple unit test would be nice (im not sure if the one you wrote can be reused)

@Atmosfearful Hmm, sorry about this. I didn't run the public api docs page to see this. Yeah, that makes sense - will move this logic over to a next api route instead.

**Comment by 0xMakka** [comment]
@Atmosfearful moved the logic out of the api into a nextjs api route instead. I didn't write a unit test for this yet, but will look to see what if I can get this done tomorrow.

File: carbonmark/lib/getPaymentMethods.ts [hunk 2]
```diff
@@ -1,23 +1,36 @@
...
     id: string;
     icon: StaticImageData;
-    label: Uppercase<CarbonmarkPaymentMethod> | "Credit Card";
+    label: Uppercase<CarbonmarkPaymentMethod> | "Credit Card" | "Bank Transfer";
     disabled: boolean;
   };
 };
 
 export const carbonmarkPaymentMethodMap: CarbonmarkPaymentMethodMap = {
   usdc: { id: "usdc", icon: USDC, label: "USDC", disabled: false },
```
Comment (biwano) [inline carbonmark/lib/getPaymentMethods.ts:17]: Why not put the minimum amounts here? This seems clean.

File: carbonmark/pages/api/bank-transfer/index.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            objectTypeId: "0-1",
+            value: first_name,
+          },
+          {
+            name: "lastname",
+            objectTypeId: "0-1",
+            value: last_name,
+          },
+          {
+            name: "who_will_this_retirement_be_credited_to_",
```
Comment (biwano) [inline carbonmark/pages/api/bank-transfer/index.ts:78]: Noob question: Why is there an underscore at the end of this?

File: carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx [hunk 4]
```diff
@@ -126,12 +132,17 @@ export const RetireForm: FC<Props> = (props) => {
...
+  const disableSubmit = () => {
+    if (!quantity || Number(quantity) <= 0) return true;
+    if (paymentMethod === "fiat") {
+      return Number(costs) < Number(fiatMinimum);
+    } else if (paymentMethod === "bank-transfer") {
+      return Number(quantity) < Number(MINIMUM_TONNE_QUANTITY_BANK_TRANSFER);
+    } else {
+      return Number(quantity) < Number(MINIMUM_TONNE_QUANTITY);
+    }
+  };
```
Comment (0xemc) [inline carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx:145]: Does this need to be a function? quantity and paymentMethod are watched show it should re-render the component when they change

File: carbonmark/lib/constants.ts [hunk 5]
```diff
@@ -140,6 +146,8 @@ export const urls = {
   docs: "https://docs.carbonmark.com",
   docsResourcesFees:
     "https://docs.carbonmark.com/get-started/understanding-fees-on-carbonmark",
+  payViaBankForm:
+    "https://api.hsforms.com/submissions/v3/integration/submit/26010207/2f87cd63-f8a7-43e9-9483-ac541a614762",
```
Comment (0xemc) [inline carbonmark/lib/constants.ts:150]: Maybe not a risk but should this be a vercel secret perhaps?

**Review by 0xemc — APPROVED** [review]
Ticked with some comments/questions

Epic work 👍

File: carbonmark/lib/getPaymentMethods.ts (hunk 2 above)
Comment (0xMakka) [reply carbonmark/lib/getPaymentMethods.ts:17]: @biwano yeah this actually seems like a great idea to keep this logic together here. I'm going to create a separate ticket to cleanup the validation logic and move it out to this file, would prefer not to start making changes to this logic now as I believe we are looking to release this feature in todays release if possible.

File: carbonmark/pages/api/bank-transfer/index.ts (hunk 3 above)
Comment (0xMakka) [reply carbonmark/pages/api/bank-transfer/index.ts:78]: @biwano So this is the id that is automatically generated in hubspot based on the title - I think it adds an underscore for any special characters, spaces etc.. - in this case the title is "Who will this retirement be credited to?", so I assume it's adding the underscore at the end for the question mark.

Screenshot 2023-12-20 at 11 48 40

File: carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx (hunk 4 above)
Comment (0xMakka) [reply carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx:145]: @0xemc valid point, I initially changed this as I found the ternary was getting complicated to read so decided to try clean it up and make it a bit more readable. Let me take another look at this though.

File: carbonmark/lib/constants.ts (hunk 5 above)
Comment (0xMakka) [reply carbonmark/lib/constants.ts:150]: @0xemc Good point, and honestly I'm not sure if there is any risk to exposing this url publicly, maybe @Atmosfearful has some thoughts on this? Happy to move it to a vercel variable if we feel it's necessary.

**Comment by jabby09** [comment]
@0xMakka did another QA pass, everything is still working as expected on the front end

**Review by Atmosfearful — APPROVED** [review]

File: carbonmark/lib/constants.ts (hunk 5 above)
Comment (Atmosfearful) [reply carbonmark/lib/constants.ts:150]: meh they are intended to be public unauthenticated urls anyways, sniffable from our site payload, only benefit of env vars would be hiding from github bots I guess🤷

**Pull request merged**